### ISSUE TITLE
ListView with sticky header and reordable items features

### ISSUE THREAD

**Issue by unveloper** [flutter/flutter#51107]
## Use case

Hi!
One effect that I'm struggling to achieve in my app is having a `ListView` that loads paginated data and has this two features:

1. sticky headers like this: https://miro.medium.com/max/346/1*74Ky84hOwhK6MNA8BvspQQ.gif
I've tried a bunch of community libraries (like https://pub.dev/packages/flutter_sticky_header or https://pub.dev/packages/sticky_headers) but none of them allow to create items on the fly like the `itemBuilder` in `ListView.builder` 

2. capability of reording items. I know that `ReorderableListView` exists but that widget accepts only a precalculated list of children and still doesn't have an "item builder" like `ListView.builder` so it's impossible to build list items on demand and having the same performance as the ListView.

## Proposal

For the part of paginated data I've created this question on StackOverflow and I'm trying to create the best implementation
https://stackoverflow.com/questions/60074466/pagination-infinite-scrolling-in-flutter-with-caching-and-realtime-invalidatio

For the part of sticky headers and reordable items having an item builder I still can't figure it out.
Is there already a widget or the tools to achieve this? If not can you please suggest a possible approach and maybe create a specific example?

Thanks

**Comment by Areopagitics** [comment]
I think ListView, GridView and CustomScrollView (all of the ScrollViews) should be able to have a sticky header; I suggested in #108289 it be done with a sticky property for [SliverPersistentHeader](https://api.flutter.dev/flutter/widgets/SliverPersistentHeader/pinned.html), I admit it's not the best, since SliverPersistentHeader is more for a resizable header (but then it should be called SliverResizableHeader). If it were added, then the concept of SliverPersistentHeader would need to change more according to the characteristic of its actual name **Persistent** and not to one that's resizable. Maybe resizable could be a property and then sticky could be added?  Just some suggestions... 

I have to admit, I haven't been the greatest fan of the Flutter nomenclature; isn't there something more standard and framework independent?

**Comment by Piinks** [comment]
This is something that would likely be done through composition. I see you have found the sticky_headers package, there is a request in https://github.com/flutter/flutter/issues/108289 for that specific feature in the framework. 
I imagine if that were added, plus the ability to group slivers like https://github.com/flutter/flutter/issues/33137 you could compose those with the SliverReorderableList... not super sure. We should close those other issues first and see if it works. This would be a great example app.

**Comment by maxstubbersfield** [comment]
Hi there, with the addition of `SliverMainAxisGroup` it is possible to do the sticky headers, however as far as I can tell this can't be done lazily as mentioned in point 1 of this issue. Is this something that will be added in future?

Thanks!

**Comment by Areopagitics** [comment]
Yes, see [here](https://github.com/flutter/flutter/pull/126596#issuecomment-1618941049) for an example

**Comment by maxstubbersfield** [comment]
> @maxstubbersfield Yes, see [here](https://github.com/flutter/flutter/pull/126596#issuecomment-1618941049) for an example

Thanks for your reply! As far as I can tell, that example isn't building the Slivers lazily/on demand. I would like to be able to generate the `SliverMainAxisGroup` when it scrolls into view. 

I am looking for something like a '_sliverBuilder_' (which currently doesn't exist afaik)
E.g.
```
CustomScrollView(
      slivers: <Widget>[
        SliverList.builder(
          sliverBuilder: (BuildContext context, int index) {
            return SliverMainAxisGroup(
              slivers: <Widget>[
                SliverPersistentHeader(
                  pinned: true,
                  floating: false,
                  delegate: Delegate(Colors.white, 'Test'),
                ),
                SliverList.builder(
                  itemBuilder: (BuildContext context, int index) {
                    return Container(
                      color: index.isEven ? Colors.amber[300] : Colors.blue[300],
                      height: 100.0,
                      child: Center(
                        child: Text(
                          'Item $index',
                          style: const TextStyle(fontSize: 24),
                        ),
                      ),
                    );
                  },
                  itemCount: 5,
                ),
              ],
            );
          },
           itemCount: 10,
        )
      ],
    );
```

**Comment by Piinks** [comment]
> I would like to be able to generate the SliverMainAxisGroup when it scrolls into view.

@maxstubbersfield do you mean SliverLayoutBuilder? 

https://api.flutter.dev/flutter/widgets/SliverLayoutBuilder-class.html

**Comment by Piinks** [comment]
Filing a new issue would be best @maxstubbersfield since this issue is not requesting that feature. Thanks!

**Comment by maxstubbersfield** [comment]
> Filing a new issue would be best @maxstubbersfield since this issue is not requesting that feature. Thanks!

I posted here because point 1 of this issue mentions the sticky headers, which is now possible without using an external package, however they also mention that they want to "create items on the fly like the itemBuilder in ListView.builder" which I understood to be the same as what I have asked about. 

Maybe I have misunderstood something here? I am happy to open another issue if that is better :)

Cheers :)
